add --discussion-category flag to release cmd

despreston opened the pull request:
Flag for signaling that a discussion should be created with the given
category for the release. Discussions are not supported for draft
releases. If a discussion category is given for a draft, an err will be
shown.

Closes #3381

<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

despreston committed:
add --discussion-category flag to release cmd

Flag for signaling that a discussion should be created with the given
category for the release. Discussions are not supported for draft
releases. If a discussion category is given for a draft, an err will be
shown.

Closes #3381

despreston commented:
Info on API for releases:
https://docs.github.com/en/rest/reference/repos#create-a-release

mislav commented on pkg/cmd/release/create/create.go:
```diff
@@ -92,9 +95,16 @@ func NewCmdCreate(f *cmdutil.Factory, runF func(*CreateOptions) error) *cobra.Co
 		`),
 		Args: cmdutil.MinimumArgs(1, "could not create: no tag name provided"),
 		RunE: func(cmd *cobra.Command, args []string) error {
+			if cmd.Flags().Changed("discussion-category") && opts.Draft {
+				return errors.New("Discussions for draft releases not supported")
```
Will the server also error out on this? If yes, I would propose removing the client-side validation. Maybe in the future the platform allows this (e.g. saves the parameter in a draft release and uses it upon publish), and in that case this bit of client-side validation will be unwanted.

despreston replied:
Just tested this. The server creates the release and ignores the discussion category. No message or anything from the server that the discussion category is not allowed for draft releases. Maybe worth relaying that to the backend team or w/e as that might be confusing, especially if you're trying to automate this stuff. I'll leave in this validation and push changes for the rest of your comments. Look forward to hearing your feedback.

mislav replied:
It's fine to leave the validation for now. I don't think it's extremely likely that the platform will accept and respect this parameter for draft releases anytime soon.

mislav commented on pkg/cmd/release/create/create.go:
```diff
@@ -48,6 +49,8 @@ type CreateOptions struct {
 	// maximum number of simultaneous uploads
 	Concurrency int
+
+	DiscussionCategoryName string
```
Nit: let's call this `DiscussionCategory` so it's not overly long!

mislav commented on pkg/cmd/release/create/create.go:
```diff
@@ -268,12 +279,20 @@ func createRun(opts *CreateOptions) error {
 		}
 	}
 
+	if opts.Draft && len(opts.DiscussionCategoryName) > 0 {
```
This validation is redundant if the `RunE` block already performs validation of cli inputs

despreston replied:
I think you'd still need to perform validation after the interactive bit on line 252, because the value of opts.Draft could change.

mislav commented on pkg/cmd/release/create/create.go:
```diff
@@ -268,12 +279,20 @@ func createRun(opts *CreateOptions) error {
+		"prerelease":               opts.Prerelease,
+		"name":                     opts.Name,
+		"body":                     opts.Body,
+		"discussion_category_name": opts.DiscussionCategoryName,
```
Should we only add this param if the flag was set? That way we avoid having to change a bunch of old tests

mislav reviewed (changes requested):
Thanks for the code and for your patience! I have a couple of minor requests.

Can't wait for this to ship so we can dogfood is from our own release automation! https://github.com/cli/cli/blob/c5371d530346ff80a848f4acb119361a12c101dd/.github/workflows/releases.yml#L162-L162

despreston committed:
pr comments

Only add discussion category to request if there is one. This eliminates
the need to update old tests.

Renaming the variable to something shorter.

mislav reviewed (approved):
It looks great. Thank you so much!
